src/api/api.jl

```diff
@@ -1,16 +1,46 @@
 module API
 
-import Libdl
+using Libdl: dlopen, dlclose, dlpath, dlsym, RTLD_LAZY, RTLD_NODELETE
 using Base: StringVector
+using Preferences: @load_preference
 
-const depsfile = joinpath(@__DIR__, "..", "..", "deps", "deps.jl")
-if isfile(depsfile)
-    include(depsfile)
+# We avoid calling Libdl.find_library to avoid possible segfault when calling
+# dlclose (#929).
+# The only difference with Libdl.find_library is that we allow custom dlopen
+# flags via the `flags` argument.
+function find_library_alt(libnames, extrapaths=String[]; flags=RTLD_LAZY)
+    for lib in libnames
+        for path in extrapaths
+            l = joinpath(path, lib)
+            p = dlopen(l, flags; throw_error=false)
+            if p !== nothing
+                dlclose(p)
+                return l
+            end
+        end
+        p = dlopen(lib, flags; throw_error=false)
+        if p !== nothing
+            dlclose(p)
+            return lib
+        end
+    end
+    return ""
+end
+
+const libpath = @load_preference("libhdf5path", nothing)
+if libpath === nothing
+    using HDF5_jll
 else
-    @error(
-        "HDF5 is not properly installed. Please run Pkg.build(\"HDF5\") ",
-        "and restart Julia."
-    )
+    libpaths = [libpath, joinpath(libpath, "lib"), joinpath(libpath, "lib64")]
+    flags = RTLD_LAZY | RTLD_NODELETE  # RTLD_NODELETE may be needed to avoid segfault (#929)
+
+    libhdf5 = find_library_alt(["libhdf5", "libhdf5_openmpi", "libhdf5_mpich"], libpaths; flags=flags)
+    libhdf5_hl = find_library_alt(["libhdf5_hl", "libhdf5_openmpi_hl", "libhdf5_mpich_hl"], libpaths; flags=flags)
+
+    isempty(libhdf5) && error("libhdf5 could not be found")
+    isempty(libhdf5_hl) && error("libhdf5_hl could not be found")
```

Comment:
This will prevent the user from loading HDF5 if the custom library has been (re)moved since it `libhdf5path` was set. IMHO (and from past experience) it would be good to add a note to the docs on how to proceed in that case.